### DISCUSSION TITLE
Storyboard Editing

### DISCUSSION THREAD

**Discussion by LeNitrous** [ppy/osu#16570] [Ideas]
Storyboard creation has been inaccessible and most if not all not use stable's in-game editor (much less so writing OSBs by hand nowadays). There are two possible paths moving forward allowing storyboard editing in lazer.

- **A more complex storyboard editor**
Basically similar to Adobe After Effects.

- **Storyboard Scripting**
Third-party community-created tools like [storybrew](https://github.com/damnae/storybrew) and [sbtw!](https://github.com/lenitrous/sbtw) which in a nutshell, allow writing storyboards with code.

With either path, it may be possible to make storyboarding more accessible and may encourage more sets to be made with storyboards.

**Comment by bdach** [comment]
don't feel confident answering this question as to which path is preferred, but just going to note that preliminary designs for a UI-based storyboard editor are already present here: https://www.figma.com/file/ytnnne2TH8Z956Jxhiypqq/Beatmap-Editor-2?node-id=1%3A1268

**Comment by Coppertine** [comment]
Just expaning on a potentionally held off idea.
I have some input that I think could be a new version of storyboards in lazer including:
- Fragment Shader support for each sprite / layer.
- Custom components that can be reused throughout sections of a beatmap.
- Ability to import / create any Drawable used in lazer (i.e. using a Visualisation drawable for your storyboard)
- Set spesific sprites to change size / position depending on mods (would support hitobject highlights and other effects that mods like Hard Rock would change)
- Create conditional triggers / make said triggers nested.
- osu! Markup Language (or similar) as the storyboard file storage method.
- "Basic" 3D implementation
- Font importing and Text generation
- "Basic" Particle generation

These are only some ideas I have as a storyboarder myself. We mostly use programming languages to create the .osb file format and have to work around extremely weird limitations of the stable storyboard feature.
I know storyboards overall are on a lower priority than stable beatmaps, but I'm only throwing ideas for what a new version of a storyboard would be. As well, I feel it's better to use this discussion instead of creating a new one.

**Reply by ching577** [reply]
I'd like to add a few features I think would be cool:

- customizable cubic bezier easing. like in CSS https://developer.mozilla.org/en-US/docs/Web/CSS/easing-function. 
- be able to use video files as sprites.

**Reply by Coppertine** [reply]
2) is already implemented in stable and lazer.

**Reply by ching577** [reply]
I mean video as in mp4, mkv, etc. not image sequences. I wanna see anime clips in storyboards >.<

**Reply by Coppertine** [reply]
videos are sprites, they can be modified at will, but unsure about lazer tho
